compiler+runtime/src/cpp/jank/codegen/llvm_processor.cpp

```diff
@@ -926,54 +964,267 @@ namespace jank::codegen
+    if(expr->finally_body.is_some())
+    {
+      landing_pad->setCleanup(true);
+    }
+
+    if(has_catch)
+    {
+      auto const object_ptr_type_info{ ctx->module->getOrInsertGlobal(
+        typeid(object_ref).name(),
+        llvm::PointerType::get(ctx->builder->getInt8Ty(), 0)) };
+      landing_pad->addClause(object_ptr_type_info);
+    }
+
+    llvm::Value *catch_val{};
+    llvm::BasicBlock *catch_end_bb{};
+
+    if(has_catch)
+    {
+      auto const cleanup_lpad_bb
+        = llvm::BasicBlock::Create(*ctx->llvm_ctx, "cleanup.lpad", current_fn);
+      lpad_and_catch_body_stack.emplace_back(cleanup_lpad_bb, nullptr);
+      util::scope_exit const pop_cleanup_lpad{ [this]() { lpad_and_catch_body_stack.pop_back(); } };
+
+      ctx->builder->CreateBr(catch_body_bb);
+      current_fn->insert(current_fn->end(), catch_body_bb);
+      ctx->builder->SetInsertPoint(catch_body_bb);
+
+      auto const &[sym, body]{ expr->catch_body.unwrap() };
+      auto old_locals(locals);
+      auto exception_ptr{ ctx->builder->CreateExtractValue(landing_pad, 0, "ex.ptr") };
+      auto const begin_catch_fn{
+        ctx->module->getOrInsertFunction("__cxa_begin_catch", ptr_ty, ptr_ty)
+      };
+      auto const caught_ptr{ ctx->builder->CreateCall(begin_catch_fn, { exception_ptr }) };
+      locals[sym] = ctx->builder->CreateLoad(ptr_ty, caught_ptr, "ex.val");
+
+      auto const original_catch_pos{ body->position };
+      body->propagate_position(expression_position::value);
+      util::scope_exit const restore_catch_pos{ [&]() {
+        body->propagate_position(original_catch_pos);
+      } };
+      catch_val = gen(body, arity);
+
+      auto const end_catch_fn{ ctx->module->getOrInsertFunction("__cxa_end_catch",
+                                                                ctx->builder->getVoidTy()) };
+      ctx->builder->CreateCall(end_catch_fn, {});
+      locals = std::move(old_locals);
+
+      /* An empty catch body will not produce a value, so `catch_val` will be nullptr. */
+      /* If there's no terminator, we must add one. */
+      if(ctx->builder->GetInsertBlock()->getTerminator() == nullptr)
+      {
+        if(!catch_val)
+        {
+          catch_val = gen_global(jank_nil);
+        }
+        catch_end_bb = ctx->builder->GetInsertBlock();
+        ctx->builder->CreateBr(finally_bb);
+      }
+
+      ctx->builder->SetInsertPoint(cleanup_lpad_bb);
+      auto const cleanup_lpad{ ctx->builder->CreateLandingPad(lpad_ty, 0) };
+      cleanup_lpad->setCleanup(true);
+      if(expr->finally_body.is_some())
+      {
+        gen(expr->finally_body.unwrap(), arity);
+      }
+      ctx->builder->CreateResume(cleanup_lpad);
+    }
+    else
+    {
+      if(expr->finally_body.is_some())
+      {
+        gen(expr->finally_body.unwrap(), arity);
+      }
+
+      if(!lpad_and_catch_body_stack.empty())
+      {
+        /* rethrow the exception to the outer catch if one is available. */
```

Comment:
Let's explain more of what's going on here and *then* lead up to what we're going to do about it (rethrow).